sklearn/decomposition/dict_learning.py

```diff
@@ -343,35 +352,32 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Updated dictionary.
 
     """
-    n_components = len(code)
-    n_samples = Y.shape[0]
+    n_features, n_components = B.shape
     random_state = check_random_state(random_state)
     # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
+    R = -np.dot(dictionary, A)
+    R += B
     R = np.asfortranarray(R)
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    ger, = linalg.get_blas_funcs(('ger',), (dictionary, A))
     for k in range(n_components):
         # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :].T)
+        R = ger(1.0, dictionary[:, k], A[k, :], a=R, overwrite_a=True)
+        dictionary[:, k] = np.dot(R, A[k, :])
```

Comment:
Confirm that your PR is faster and reaches the same error.

With tol=1e-20 the number of iterations needed for convergence fluctuates (both on master and here), but I think (although odd) it's a numerical issue. With reasonable tol 1e-6 the number of iterations is exactly the same. The returned U and V have the same norm as well, so I think you're right in saying the paths are identical.
master
```
[...]
Iteration  166 (elapsed time:   3s,  0.1mn, current cost  198.781)
--- Convergence reached after 166 iterations
198.780858248
Time elapsed: 3.13
U norm 14.4716684728
V norm 1.73205080757

```
here
```
[...]
Iteration  150 (elapsed time:   1s,  0.0mn)
--- Convergence reached after 166 iterations
Learning code... done (total time:   1s,  0.0mn)
198.780765235
Time elapsed: 1.99
U norm 14.4722062044
V norm 1.73205080757
```